### ISSUE TITLE
Type infer issue in interfaces with internal properties

### ISSUE THREAD

**Issue by fuson** [microsoft/TypeScript#42716]
# Bug Report

### 🕗 Version & Regression Information
4.1.3 & 4.2.0-dev the same

### ⏯ Playground Link
https://www.typescriptlang.org/play?ts=4.2.0-beta#code/JYOwLgpgTgZghgYwgAgEJwM4oN4ChkHKgbAAmEAXMnobcgG5UgCuAtgEbQDc+hAvrgG5QkWIhQANZBAAekEKQxpMOXgWJlK1NXUbIAjDoG04TNpyg9aMgPxmO3QbmHho8JMgCa0uRAVL0LG1aDXIqGjoCPQAmIx1TZBYHSx0ATztE80chETdxZAAtH3lFZSCI9RASMODIhioAZjjaAC8MpIseIQQAeyqwBmj7C2QAXmRongAbCAH6fSp9AB9J52AYZAAKemix0fHogEoGfTGJrmkpoPmz-R5cXv6GOCnmLSklrzOKoirNcJ0tD08yMABp4osAAyCe6PDA9GYAOimPQA5tsXm9DjwgA

### 💻 Code

```ts
interface Base {
    inside: {
        v: number;
    }
}

interface X extends Base {
    inside: {
        v: 1
    }
    a: number;
    x?: number;
}

interface Y extends Base {
    inside: {
        v: 2
    }
    a: number;
    y?: number;
}

interface Z extends Base {
    inside: {
        v: 3
    }
    z?: number;
}

const v2: number = 2;
let v1: 1|2;

if (v2 === 2) v1 = 2; else v1 = 1;

const value: X | Y = {
    inside: {
        v: v1
    },
    a: 10
};

console.log(value);
```

### 🙁 Actual behavior

Type '{ inside: { v: 1 | 2; }; a: number; }' is not assignable to type 'X | Y'.
  Type '{ inside: { v: 1 | 2; }; a: number; }' is not assignable to type 'Y'.
    Types of property 'inside' are incompatible.
      Type '{ v: 1 | 2; }' is not assignable to type '{ v: 2; }'.(2322)

### 🙂 Expected behavior

No error ^^

**Comment by RyanCavanaugh** [comment]
@rbuckton I would have expected our union-combinatorial thing to kick in here, maybe I missed something?

**Comment by fuson** [comment]
Any news about this bug?

**Comment by danvk** [comment]
I believe this is the same issue, just with slightly less code:

```ts
interface A { a: string; }
interface B { b: string; }
interface AWrap { test: A; }
interface BWrap { test: B; }

declare let p: A | B;
const pw: AWrap | BWrap = { test: p };
//    ~~
// Type '{ test: A | B; }' is not assignable to type 'AWrap | BWrap'.
//   Type '{ test: A | B; }' is not assignable to type 'BWrap'.
//     Types of property 'test' are incompatible.
//       Type 'A | B' is not assignable to type 'B'.
//         Property 'b' is missing in type 'A' but required in type 'B'. (2322)
```

([Playground link][1])

Unless I'm missing something, that assignment should be safe. cc @ludazhao

This is the only workaround that I came up with, it definitely looks a bit strange!

```ts
function isA(x: A | B): x is A {
  return 'a' in x;
}

const pw : AWrap | BWrap = isA(p) ? { test: p } : { test: p };  // no error
```

[1]: https://www.typescriptlang.org/play?ts=4.4.0-beta#code/JYOwLgpgTgZghgYwgAgILIN7LgLmQZzClAHMBuZAXwChRJZEUAhTZAIz0OJHKtvGjwkaAOpQ4AB1aRCeVBRp1BjZEzGTpEWaoXVqAEwgIANnCgpjEMMglzkAHx3UEAexCEbAdznqpjteJSALya2lKUZNQA9FHIccgAfgnRsQAqAJ4SKADkWDJgdv4K2cjA+MggLtZw+PjAJCBwbJbIYC6tmTmovg6qvtkAdCnxGVnIua1aBWi9TMWl5ZXVtfWNzShtHWPZAZKDw-HIo1rILjA2UC5ZUGDp4-klZiigrgC2EnBgwOtDMYcjnXG6H8JTKFSq2BWDSaLU2t22TH2f3+8QACpdrrdxmxQeVXmU6jxSiAtl0SmwAK7WcwARwpwHM+mJpPGiIGyAAFAAmADMXK5AEo9NEAFTUGAUkAIL5uBaoDkAD0KqgFeAVCxmGGocXMYApUBJAEJDQBZT4ACwG4hA+hcrw5AsiNGcbg8Ek8yB8gVmPRCZXlcC1cXyeH9HIkAuQAH4bJ6bAAaagRagiqJAA

**Comment by fuson** [comment]
@RyanCavanaugh any chance to fix? :)

**Comment by rbuckton** [comment]
I don't believe we consider `inside` to be a discriminant property since it is not itself a literal type. The union-combinatorial assignability mechanism depends on the constituents being discriminable.
